src/java.base/share/classes/jdk/internal/loader/BuiltinClassLoader.java

```diff
@@ -1085,4 +1085,10 @@ private void resetArchivedStates() {
         ucp = null;
         resourceCache = null;
     }
+
+    // Called during -Xshare:dump from AppClassLoader.resetArchivedStates().
+    void resetArchivedStatesForAppClassLoader() {
+        setClassPath(null);
+        if (!moduleToReader.isEmpty()) moduleToReader.clear();
```

Comment:
```suggestion
        if (!moduleToReader.isEmpty()) {
            moduleToReader.clear();
        }
```

Also, do we need to do the same thing for the platform loader as well?

Reply:
Added braces.
The `setClassPath(null)` used to be in `ClassLoaders.AppClassLoader`. Based on investigations so far, the clearing of the `moduleToReader` map is required only for `AppClassLoader`.

Reply:
Why does it need to clear `moduleToReader` only for app loader and not for platform loader? Is it because the `moduleToReader` for the app loader may contain reference to jar files that indirectly references some file system objects?

Since moduleToReader is just a cache, I think it's better to always clear it for both loaders. Also, the logic can be moved into BuiltinClassLoader: 

```
class BuiltinClassLoader {
        ....
        private void resetArchivedStates() {
            setClassPath(null);
            moduleToReader.clear();
        }
```

Note that `reset_states()` in heapShared.cpp walks the class hierarchy and calls `resetArchivedStates()` for platform/app class loaders and all of their super classes.

Reply:
setClassPath(null) is the same as `ucp = null` but yes, keep it simple as otherwise there will be question each time there are changes. BuiltinClassPath should not include any code that is specific to the app class loader or the platform class loader as there are specific subclasses for that.

Reply:
Ok. I've simplified the fix as suggested.